### ISSUE TITLE
[Feature Request] Seed id hash with parent's id

### ISSUE THREAD

**Issue by bullno1** [nicbarker/clay#26]
Currently, element id needs to be globally unique.
Copy&paste or refactoring reusable components into functions would be a bit of a pain since you'd need to parameterize all the ids.

What if the id only has to be unique among siblings?
`Clay__Rehash` already exists but it is only used for debugger.
The fact that it exists mean that kind of id is useful.

`CLAY_ID_AUTO` on the other hand, does not allow providing a string and counter which are needed to identify elements independent of orders.
e.g: In a list of items.

There are 2 ways of doing this: 

* Make `CLAY_ID` and `CLAY_IDI` context-dependent.
   This might break some existing code.

   Floating attachment may be a bit harder but there are workarounds like saving the id into a variable first.
* Alternatively, provide `CLAY_LOCAL_ID` and `CLAY_LOCAL_IDI` which takes the parent id as [seed](https://github.com/nicbarker/clay/blob/main/clay.h#L1360).
   Alternate names:
   
   * `CLAY_CTX_ID[I]`: context dependent id
   * `CLAY_SCOPED_ID[I]`

I can make PR for either of these.

**Comment by nicbarker** [comment]
Hello! Firstly, I completely agree with your assertion that we need a local context-aware way to auto generate IDs. At present it's quite painful to have to uniquely everything in the hierarchy, even for immediate mode renderers where it's almost entirely unnecessary. I'll merge your PR in the short term to address that particular problem.

Secondly given that you've dug in here already, I'd like to get your opinion on an API change I'm working on at the moment. It replaces all the type-specific element macros with a single generic `CLAY()` macro, and allows you to configure this element by attaching whichever configs you want:

![carbon(88)](https://github.com/user-attachments/assets/89ece383-e33b-4a36-8f30-f84c2353a6c4)

I specifically started this change to address the painful ergonomics of working on an interface where most elements needed a border _and_ a rectangle.

The main idea is that in the new api, _all_ configs are optional, _including ID and Layout_, e.g:

```C
// Create an empty layout element with no children and the default layout config
CLAY()

// Create a scrolling element with a border and background
CLAY(CLAY_SCROLL(...etc), CLAY_BORDER(...etc), CLAY_RECTANGLE(...etc)) {
    // ...children
}

// Tag a rectangle with a specific ID for mouse interactions, etc
CLAY(CLAY_ID("Button"), CLAY_RECTANGLE(...etc)) {
    // ...children
}
```
With the refactor being relevant to this situation because I'm planning to attach generated ids to each element if the user chooses not to provide one. It makes sense for the auto generated ID to be derived from the parent ID, and positional based on the sibling index.

**Comment by nicbarker** [comment]
This was addressed in https://github.com/nicbarker/clay/pull/27 but I will leave it open until I've updated the `README` to include documentation of the new feature.

**Comment by bullno1** [comment]
> It replaces all the type-specific element macros with a single generic CLAY() macro, and allows you to configure this element by attaching whichever configs you want

Since the user can pick a combination of features, how about having just one type: `Clay_Element`?
Then features can be turned on by initializing it.

```c
CLAY(
    CLAY_SCROLL(...), // Enable scrolls
    CLAY_BORDER(...), // Enable border
) {
}
```

The API is essentially the same but the implementation is different.
There is only `Clay_Element`.
The renderer just have to check whether a feature is enabled like:

```c
if (cmd.has_border) {
    // draw border
}

if (cmd.has_image) {
    // draw image background
}

```

**Comment by nicbarker** [comment]
Yep, you're describing pretty much my exact line of thinking - a single macro that can be configured with whichever features. There is a little internal work to do to support this but it shouldn't take too long.

> Another radical idea about id: Given that local id seems to be more prevalent, how about make it the default (CLAY_ID is local)? And have CLAY_ID_GLOBAL to create global id instead?

Yes, I'm very much thinking along the same lines. I want it to be obvious when an element is tagged with a global ID for a reason - it's used for click handling etc.

I actually would prefer if the IDs were local internally, but mostly auto generated. e.g. if you don't specify an ID at all:

```C
CLAY_ELEMENT(CLAY_SCROLL(), CLAY_RECTANGLE()) { // Automatically assigned CLAY_IDI_LOCAL("0", 0)
    CLAY_ELEMENT() // Automatically assigned CLAY_IDI_LOCAL("0", 0)
    CLAY_ELEMENT() // Automatically assigned CLAY_IDI_LOCAL("0", 1)
    CLAY_ELEMENT() // Automatically assigned CLAY_IDI_LOCAL("0", 2)
    CLAY_ELEMENT() // Automatically assigned CLAY_IDI_LOCAL("0", 3)
    CLAY_ELEMENT(CLAY_TAG("ClickableButton")) {} // ID is already specified with a global so no local ID is attached
}
```

**Comment by nicbarker** [comment]
😁  Finally wrote the README section for this here: https://github.com/nicbarker/clay/commit/ebeef93c34a393813aaf789f76149ab1f5502bcd
